### ISSUE TITLE
GPU kernel nodes (sgemm, NCCL) have no outgoing dependencies in chakra_trace_linker output

### ISSUE THREAD

**Issue by XavierQuerol** [mlcommons/chakra#215]
## Describe the Bug
I am running a simple single-GPU GPT training script with the PyTorch profiler to collect execution traces, which I then convert to Chakra ET format using `chakra_trace_link` and `chakra_converter`. 

After inspecting the final `chakra_read_rank0.json`, I find that **GPU kernel nodes (sgemm, elementwise kernels, etc.) are never listed as `dataDeps` or `ctrlDeps` of any other node in the graph.** Every GPU kernel acts as a leaf. The only edges that exist go from CPU ops to their GPU kernel children—never in the other direction, and never between GPU kernels.

The same behavior occurs with **NCCL collective operations**. In practice, this causes incorrect behavior in simulators: because NCCL kernel nodes have no important incoming dependencies, they are immediately free to execute at the very start of the simulation. In a real training run, they should only start during the backpropagation phase after the relevant gradients have been computed (as in the real run)

I am not sure whether the problem originates in the `chakra_trace_link` step, the `chakra_converter` step, or in how the input traces are collected. However, I do not think it is a profiling issue, as the raw Kineto and PyTorch ET files appear to contain the expected events and correlation IDs.

---

### 1. Collect Traces
The relevant profiler setup in the training script is configured as follows:

```python
from torch.profiler import ExecutionTraceObserver, profile
from torch.autograd.profiler import _ExperimentalConfig

et = ExecutionTraceObserver()
et.register_callback(pyet_path)

with profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ],
    schedule=torch.profiler.schedule(
        skip_first=2, wait=1, warmup=3, active=1, repeat=1
    ),
    record_shapes=True,
    with_flops=True,
    on_trace_ready=trace_handler,
    execution_trace_observer=et,
    experimental_config=_ExperimentalConfig(enable_cuda_sync_events=True),
) as prof:
    for step, batch in enumerate(train_dataloader):
        # ... forward / backward / optimizer step ...
        prof.step()

et.unregister_callback()
```
This produces `pytorch_et_rank0.json` (schema `1.1.0-chakra.0.0.4`) and `kineto_trace_rank0.json`.

### 2. Link and Convert
Run the following post-processing pipeline:

```bash
chakra_trace_link \
    --rank 0 \
    --chakra-host-trace ./pytorch_et_rank0.json \
    --chakra-device-trace ./kineto_trace_rank0.json \
    --output-file ./merged/merged_rank0.json

chakra_converter PyTorch \
    --input ./merged/merged_rank0.json \
    --output ./chakra/chakra_rank.0.et --simulate

chakra_jsonizer \
    --input_filename  ./chakra/chakra_rank.0.et \
    --output_filename ./chakra_read/chakra_read_rank0.json
```

**Environment & Version Information:**
* **Chakra:** 1.0.0 (via pip)
* **PyTorch:** 2.7 (Schema: `1.1.0-chakra.0.0.4`)

### 3. Inspect the Output
In `chakra_read_rank0.json`, taking three representative nodes as an example:
* **Node 361 (`aten::mm`):** CPU op, 59 µs, `dataDeps: [358, ...]`
* **Node 362 (`ampere_sgemm_128x32_sliced1x4_tn`):** GPU kernel, 98078 µs, `dataDeps: [361]`, `ctrlDeps: [361]`
* **Node 372 (`ampere_sgemm_64x64_tn`):** GPU kernel, next sgemm in the trace, `dataDeps: [<its own CPU launcher>]`

Node 362 **does not appear** in the `dataDeps` or `ctrlDeps` of any other node. Searching the entire JSON file for `"362"` as a dependency value returns zero results.

---

## Expected Behavior
I would expect some dependency path to exist between consecutive GPU kernels (such as node 362 and node 372). As it stands, all GPU kernel nodes become dependency-free simultaneously once their single CPU-launcher edge is satisfied. This leaves simulators with no timing info to prevent issuing them all at once, which does not reflect real execution.

Two possible architectural approaches come to mind, though I am unsure which is intended:
1. **Intra-stream chaining:** All kernels submitted to the same CUDA stream are linked sequentially (`kernel[i+1]` depends on `kernel[i]`), since CUDA guarantees stream-ordered execution.
2. **GPU kernel promoted into the CPU chain:** Any node that currently has a `dataDep` on a CPU launcher (e.g., node 363 depending on 361) would instead depend on that launcher's GPU child (e.g., node 363 depends on 362). This places the GPU kernel in the middle of the execution chain rather than leaving it dangling as a leaf.

---

## Attachments
The reduced input and output trace files are attached below for reference:

* [chakra_read_rank0.json](https://github.com/user-attachments/files/28309677/chakra_read_rank0.json)
* [pytorch_et_rank_reduced.json](https://github.com/user-attachments/files/28309678/pytorch_et_rank_reduced.json)
* [kineto_trace_reduced.json](https://github.com/user-attachments/files/28309676/kineto_trace_reduced.json)

**Comment by XMaroRadoX** [comment]
Hi Xavier, thanks for documenting this. Our project at Politecnico di Torino studies congestion control for AI clusters, and we ran into the same failure while sending PyTorch/Kineto traces through Chakra and AstraSim.

In our captures, compute and NCCL kernels depended on their CPU launchers but had no successors on the GPU stream. The simulator could therefore issue collective work before the corresponding training phase. We caught it because the simulated ordering contradicted both the Kineto timeline and the tensor producer/consumer relationships. That made the affected timing and congestion results unusable until we rebuilt the missing causality.

We first repaired the ordering in our downstream research pipeline and used independent checks to requalify our traces. The linked trace already carries the evidence needed for an upstream fix: device ID, stream ID, timestamp, and explicit synchronization records. We have now submitted #221 with the smallest version of that repair in Chakra itself. It preserves launcher dependencies, chains operations only within the same device and stream, uses input order for timestamp ties, and keeps explicit cross-stream synchronization separate. It adds no schema dependency on our project.

We ran Chakra's full Python test suite through SLURM: 143 tests pass, including the new stream, collective, copy, synchronization, missing-metadata, duplicate-edge, and cycle cases. We also reconstructed a converter regression from the four kernel records attached here. The original attachments are JSON fragments, so we describe that test as a reconstructed converter case rather than a full linker reproduction.

Your report saved us time when we traced the fault across the pipeline. Feedback on #221 would be welcome, especially if you still have the complete reduced linked JSON and can run it against the branch.
